FIX: Bug in peak_finder() and find_eog_events() #5025

TommyClausner opened the pull request:
Fixes #5025 
Problems resulted mostly from data type handling. I added  two
statements checking whether the output type is as desired (i.e. ndarray
of dtype int64 or float64). Further, problems raised, when the input
vector was empty. Added a check to the check for dimensionality.

agramfort commented on mne/preprocessing/peak_finder.py:
```diff
@@ -161,6 +160,13 @@ def peak_finder(x0, thresh=None, extrema=1, verbose=None):
         x0 = -x0
 
+    # ensure output type array
+    if type(peak_inds) is not np.ndarray:
```
use isinstance

agramfort commented:
There was already a test file in preprocessing/tests folder. Update this one

larsoner commented on mne/tests/test_peak_finder.py:
```diff
@@ -0,0 +1,35 @@
+    assert_equal(peak_inds.dtype, np.dtype('int64'))
+    assert_equal(peak_mags.dtype, np.dtype('float64'))
+
+run_tests_if_main()
```
please remove this file `git rm`

agramfort commented on mne/preprocessing/tests/test_peak_finder.py:
```diff
@@ -1,10 +1,38 @@
 
 
 def test_peak_finder():
-    """Test the peak detection method"""
```
don't remove this docstring headline please

agramfort commented on mne/preprocessing/tests/test_peak_finder.py:
```diff
@@ -1,10 +1,38 @@
-    peak_inds, peak_mags = peak_finder(x)
+
+    # check for random data
+    peak_inds, peak_mags = peak_finder(np.asarray(np.random.random(20)))
```
you don't need the asarray here as np.random.random already returns an array

agramfort commented on mne/preprocessing/tests/test_peak_finder.py:
```diff
@@ -1,10 +1,38 @@
+    assert_equal(peak_mags.dtype, np.dtype('float64'))
+
+    # check for no peaks
+    peak_inds, peak_mags = peak_finder(np.asarray(np.zeros(20)))
```
you don't need the asarray here as np.zeros already returns an array

agramfort commented on mne/preprocessing/tests/test_peak_finder.py:
```diff
@@ -1,10 +1,40 @@
-    peak_inds, peak_mags = peak_finder(x)
+
+    # check for random data
+    peak_inds, peak_mags = peak_finder(np.random.random(20))
```
we tend to fix the seeds to avoid random bugs. So it should be:

```
rng = np.random.RandomState(42)
peak_inds, peak_mags = peak_finder(rng.random(20))
```

TommyClausner replied:
The proposed command yields an error (rng does not have a module called random). So I just defined the random seed an kept the random number creation untouched.

agramfort replied:
you missed the other half :)

agramfort replied:
try rng.randn

TommyClausner replied:
:P - I know that randn works - but is it MNE-ish to create the rng object, because the other way you'd save an object

TommyClausner replied:
Didn't feel like I would be in the right position to complain :D - so I changed it to the rng...

TommyClausner commented:
just adding this line in Changelog:
- Fixed output datatype in :func:`mne.preprocessing.peak_finder` and added empty input check by `Tommy Clausner`

TommyClausner commented:
(sorry meant the Bug section of course ;) )

agramfort commented:
yes

make sure link works

agramfort commented on doc/whats_new.rst:
```diff
@@ -84,6 +84,8 @@ Changelog
 Bug
 ~~~
 
+- Fix bug in :func:`mne.preprocessing.peak_finder` to output datatype consistently and added input check for empty vectors by `Tommy Clausner`
```
`Tommy Clausner`_

with _ otherwise link won't work

agramfort commented:
thx @TommyClausner !
